tests/phpunit/CRM/Contribute/ActionMapping/ByTypeTest.php

```diff
@@ -319,6 +321,22 @@ public function testTokenRendering(): void {
     ];
     $this->mut->checkMailLog($expected);
 
+    $tokenProcessor = new TokenProcessor(\Civi::dispatcher(), [
+      'controller' => get_class(),
+      'smarty' => FALSE,
+      'schema' => ['contributionId'],
```

Comment:
This 'schema' is required if I copy the mechanism from https://github.com/civicrm/civicrm-core/pull/21144 - but that feels like it should be unnecessary

Reply:
Off the cuff, I'm OK with that because (one way or another) the signal has to be provided in the `listTokens` use-case, and it's good discipline to have an upfront idea of what data is going in.

OTOH, yes, it's sort of redundant for this use-case. There's enough information in the row-data to infer that `contributionId` is present. I suppose an opportune spot to leverage that might be at the start of `TokenProcessor->evaluate()` -- recompute `schema` to include `array_unique(array_merge(array_keys(...each row...))`. And the recently added bit in `ActionSchedule::sendMailings()` (vis-a-vis `schema`) would be -3 SLOC if the autodetection were better.